---
<br>
<br>
<font size="100px" face="Impact"><center>
Pandas
</center></font>
<br>
<font size="2px" face="Georgia"><center>
(훌륭한 자료를 남겨주신 선배님들께 감사드립니다.)
</center></font>





<br>
<br>

<img src="img/판다스를 집에 들이셔야 합니다.jpg" width=300></img>

<br>

<font size="2px" face="Georgia"><center>
*"이 판다스를 집에 들이십시오, 어머니"*
<br>
    *"예서가 데이터 분석에 성공할 가장 효과적인 패키지입니다"*
</center></font>



---
<font size='10px' face="Impact">
Index</font>
<br>
- <font face="Georgia">Pandas 소개</font><br>
- <font face="Georgia">Data 불러오기</font><br>
- <font face="Georgia">DataFrame 살펴보기</font><br>
- <font face="Georgia">Columns 생성, 수정, 삭제</font><br>
- <font face="Georgia">Indexing & Slicing</font><br>
- <font face="Georgia">Groupby</font><br>

<br>



# 0. Pandas 소개

---
<br>
- **판다스? 팬더스? 뭐지...**

**Pandas : '구조화된 데이터'의 처리를 지원하는 Python 라이브러리**<br>
<br>
쉽게 말하자면 엑셀과 비슷합니다.<br>
판다스를 까먹었다고 간단한 전처리는 그냥 엑셀로 해버리는 사람도 봤어요; <br>
그런데 제가 알기로는 엑셀이 불러올 수 있는 최대 row의 수가 20만도 채 안되는 걸로 알고 있습니다. 여는 데에 시간도 오래 걸리구요.<br>
**사이즈가 큰 데이터를 빠르게 처리**할 수 있는 오픈소스, 바로 판다스입니다.<br>
궁금하신 분은 아래 홈페이지도 들어가보세요. (별거 없긴 합니다...)<br>
http://pandas.pydata.org/index.html<br>
<br>
- **그래서 판다스를 왜 씀?**

**Pandas는 말하자면 Numpy의 Wrapper입니다.**<br>
Pandas는 고성능 Array 계산 라이브러리인 Numpy와 통합, 강력한 스프레드시트 처리 기능을 제공합니다.<br>
<br>
- **넘파이? 눔피? 는 또 뭐지...**

**Numpy는 Numerical Python의 약어로, 파이썬의 고성능 과학 계산용 패키지입니다.** 사실상 Array 연산의 표준이라고 볼 수 있습니다.<br>
특징으로는 일반 list에 비해 빠르고, 보다 메모리 효율적이며, **반복문 없이** 데이터 배열에 대한 처리를 지원한다는 점 등이 있습니다.<br>
low-level 언어와의 호환(통합)도 가능하다고 하네요.<br>
<br>
- **좀 더 설명을...**

<img src="http://jakevdp.github.io/images/array_vs_list.png"></img>

**결론부터 말하자면 Python list보다 array가 더 빠른 연산이 가능하고, 그렇기에 numpy와 pandas를 사용하는 것입니다.**<br>
위 도식을 보면 아시겠지만, 넘파이의 레이아웃이 저장 용량이나 접근성 측면에서 훨씬 효율적인 것을 알 수 있습니다.<br>
그렇다고 list가 절대적으로 열등한 형식이란 말은 아닙니다. list는 array보다 훨씬 유연하고, 제공되는 메서드와 함수도 다양합니다.<br>
array는 데이터 타입도 하나만 넣을 수 있는 데에 비해 list는 다양한 데이터 타입의 엘리먼트를 포함할 수 있구요.<br> 따라서 일단 둘은 목적에 따라 쓰임새가 다르다는 것 정도만 이해하시고, 더 자세한 내용은 컴과 학생에게 물어보거나 아래 링크를 참고하세요.<br>
https://jakevdp.github.io/blog/2014/05/09/why-python-is-slow/


<br>
<br>
## Introduction (Series, DataFrame)
---

Series, DataFrame의 개념을 이해하고 그 변환을 다룹니다.

- <font color ="#ff5b00">Series()</font>
- <font color ="#ff5b00">pd.DataFrame()</font>



In [1]:
import numpy as np
import pandas as pd

In [2]:
# Series는 라벨링이 되어있는 Array

list_data = [1,2,3,4,5]
list_name = ["a","b","c","d","e"]
example_obj = pd.Series(data = list_data, index=list_name)
example_obj

a    1
b    2
c    3
d    4
e    5
dtype: int64

In [3]:
example_obj.index

Index(['a', 'b', 'c', 'd', 'e'], dtype='object')

In [4]:
example_obj.values

array([1, 2, 3, 4, 5])

In [5]:
# 딕셔너리 형테의 자료도 변환 가능

dict_data = {"a":1, "b":2, "c":3, "d":4, "e":5}
example_obj = pd.Series(dict_data, dtype=np.float32, name="example_data")
example_obj

a    1.0
b    2.0
c    3.0
d    4.0
e    5.0
Name: example_data, dtype: float32

In [6]:
# 딕셔너리 형테의 자료를 데이터프레임으로 변환

raw_data = {'first_name': ['Jason', 'Molly', 'Tina', 'Jake', 'Amy'],
        'last_name': ['Miller', 'Jacobson', 'Ali', 'Milner', 'Cooze'],
        'age': [42, 52, 36, 24, 73],
        'city': ['San Francisco', 'Baltimore', 'Miami', 'Douglas', 'Boston']}
df = pd.DataFrame(raw_data, columns = ['first_name', 'last_name', 'age', 'city'])
df

first_name last_name  age           city
0      Jason    Miller   42  San Francisco
1      Molly  Jacobson   52      Baltimore
2       Tina       Ali   36          Miami
3       Jake    Milner   24        Douglas
4        Amy     Cooze   73         Boston

<br>
<br>
## numpy > list comprehension > for loop
---

for loop과 numpy의 연산 시간이 꽤 차이나는 것을 확인할 수 있습니다.<br>
pandas에서 Series에 for loop을 쓴다면 아마 영원히 없어지지 않는 * 를 보게 될 것입니다.<br>
웬만한 작업은 pandas의 method, 특히 groupby로 다 처리할 수 있습니다.<br>
<br>
<br>

In [7]:
def sclar_vector_product(scalar, vector):
    result = []
    for value in vector:
        result.append(scalar * value)
    return result 

iternation_max = 100000

vector = list(range(iternation_max))
scalar = 2

%timeit sclar_vector_product(scalar, vector) # for loop을 이용한 성능

9.65 ms ± 201 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [8]:
%timeit [scalar * value for value in range(iternation_max)] # list comprehension을 이용한 성능
%timeit np.arange(iternation_max) * scalar # numpy를 이용한 성능

8.14 ms ± 111 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
76.2 µs ± 824 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)



# 1. Data 불러오기
---

일단 pandas를 import합니다.내장함수를 제외한 라이브러리를 사용할 때는 아래와 같이 import해줘야 합니다.

In [9]:
import pandas as pd
import numpy as np

항상 pandas를 풀네임으로 다 적기가 귀찮습니다. pd로 불러오는 것이 훨씬 편하겠죠.<br>
Pandas의 pd, numpy의 np는 모든 사람들이 암묵적으로 합의한 alias(별칭)이니 웬만하면 위와 같이 import하여 사용하도록 합시다.<br>

---

#### * DataFrame 불러오기
그럼 이제, 바로 데이터를 불러와봅시다. 데이터의 종류에 따라 pd.read_{확장자}라는 method를 사용하면 됩니다.
- csv : <font color ="#ff5b00">pd.read_csv</font>
- excel : pd.read_excel
- sql : pd.read_sql
<br>
<br>
우리는 주로 csv(comma separated value)형태의 데이터를 사용하게 됩니다.
<br>
엑셀 파일은 어떻게 불러오는지 물어보는 분들이 가끔 계십니다. 저 read_excel을 잘 기억해두시거나, 그냥 excel 파일을 csv로 변환해서 쓰세요.

---

<img src="img/귀여운 피카츄.png" width=200></img>

####  오늘 실습에 사용할 데이터는 귀여운 포켓몬 600여 종의 정보가 담긴 pokemon.csv입니다! 
csv 형태의 파일을 불러와서 pokemon이라는 변수에 저장해봅시다.



In [10]:
pokemon = pd.read_csv("data/pokemon.csv")

In [11]:
pokemon.head()

Number        Name Type_1  Type_2  Total  HP  Attack  Defense  Speed  \
0       1   Bulbasaur  Grass  Poison    318  45      49       49     45   
1       2     Ivysaur  Grass  Poison    405  60      62       63     60   
2       3    Venusaur  Grass  Poison    525  80      82       83     80   
3       4  Charmander   Fire     NaN    309  39      52       43     65   
4       5  Charmeleon   Fire     NaN    405  58      64       58     80   

   Generation  isLegendary  Color  hasGender  Height_m  Weight_kg  Catch_Rate  \
0           1        False  Green       True      0.71        6.9          45   
1           1        False  Green       True      0.99       13.0          45   
2           1        False  Green       True      2.01      100.0          45   
3           1        False    Red       True      0.61        8.5          45   
4           1        False    Red       True      1.09       19.0          45   

       Body_Style  
0       quadruped  
1       quadruped  
2       quadruped  
3  bipedal_tailed  
4  bipedal_tailed

---
#### * read_csv 세부 설정
- header : 데이터에 column의 label이 있는 경우 default, 없을 경우 None으로 설정
- nrows : 불러올 행의 수. 지정하지 않을 경우 전체 데이터를 가져온다. (자주 쓰진 않아요. 데이터가 너무 클 때 적당히 끊는 용도로 씁니다.)
- usecols : 불러올 column을 정합니다. nrows와 마찬가지로 큰 데이터의 일부를 읽을 때 사용됩니다.
- index_col : index로 설정할 column. 지정하지 않으면 0부터의 array로 새로 생성됨. 이건 자주 쓰니 꼭 알아두세요!
- encoding : 운영체제나 파일 형식에 따라 encoding error가 발생할 수 있습니다. 오류가 날 때만 'euc-kr'을 시도해보세요. (한글을 위한 인코딩에는 'UTF-8'과 'EUC-KR'이 있습니다.)

In [12]:
# Header가 없는 데이터의 경우에는 header = None을 사용
# pokemon.csv는 header가 있지만 header = None을 한번 해봅시더
pd.read_csv("data/pokemon.csv", header = None, nrows = 5)

0           1       2       3      4   5       6        7      8   \
0  Number        Name  Type_1  Type_2  Total  HP  Attack  Defense  Speed   
1       1   Bulbasaur   Grass  Poison    318  45      49       49     45   
2       2     Ivysaur   Grass  Poison    405  60      62       63     60   
3       3    Venusaur   Grass  Poison    525  80      82       83     80   
4       4  Charmander    Fire     NaN    309  39      52       43     65   

           9            10     11         12        13         14          15  \
0  Generation  isLegendary  Color  hasGender  Height_m  Weight_kg  Catch_Rate   
1           1        False  Green       True      0.71        6.9          45   
2           1        False  Green       True      0.99       13.0          45   
3           1        False  Green       True      2.01      100.0          45   
4           1        False    Red       True      0.61        8.5          45   

               16  
0      Body_Style  
1       quadruped  
2       quadruped  
3       quadruped  
4  bipedal_tailed

In [13]:
# usecol에는 항상 index_col이 포함되어 있어야 합니다.

pd.read_csv("data/pokemon.csv",
            index_col = 'Name',
            nrows = 5,
            usecols = ['Name','Number','Attack','Defense'])

Number  Attack  Defense
Name                               
Bulbasaur        1      49       49
Ivysaur          2      62       63
Venusaur         3      82       83
Charmander       4      52       43
Charmeleon       5      64       58

In [14]:
pokemon = pd.read_csv("data/pokemon.csv", index_col = 'Name')

In [15]:
pokemon.head(20)

Number  Type_1  Type_2  Total  HP  Attack  Defense  Speed  \
Name                                                                    
Bulbasaur        1   Grass  Poison    318  45      49       49     45   
Ivysaur          2   Grass  Poison    405  60      62       63     60   
Venusaur         3   Grass  Poison    525  80      82       83     80   
Charmander       4    Fire     NaN    309  39      52       43     65   
Charmeleon       5    Fire     NaN    405  58      64       58     80   
Charizard        6    Fire  Flying    534  78      84       78    100   
Squirtle         7   Water     NaN    314  44      48       65     43   
Wartortle        8   Water     NaN    405  59      63       80     58   
Blastoise        9   Water     NaN    530  79      83      100     78   
Caterpie        10     Bug     NaN    195  45      30       35     45   
Metapod         11     Bug     NaN    205  50      20       55     30   
Butterfree      12     Bug  Flying    395  60      45       50     70   
Weedle          13     Bug  Poison    195  40      35       30     50   
Kakuna          14     Bug  Poison    205  45      25       50     35   
Beedrill        15     Bug  Poison    395  65      90       40     75   
Pidgey          16  Normal  Flying    251  40      45       40     56   
Pidgeotto       17  Normal  Flying    349  63      60       55     71   
Pidgeot         18  Normal  Flying    479  83      80       75    101   
Rattata         19  Normal     NaN    253  30      56       35     72   
Raticate        20  Normal     NaN    413  55      81       60     97   

            Generation  isLegendary   Color  hasGender  Height_m  Weight_kg  \
Name                                                                          
Bulbasaur            1        False   Green       True      0.71        6.9   
Ivysaur              1        False   Green       True      0.99       13.0   
Venusaur             1        False   Green       True      2.01      100.0   
Charmander           1        False     Red       True      0.61        8.5   
Charmeleon           1        False     Red       True      1.09       19.0   
Charizard            1        False     Red       True      1.70       90.5   
Squirtle             1        False    Blue       True      0.51        9.0   
Wartortle            1        False    Blue       True      0.99       22.5   
Blastoise            1        False    Blue       True      1.60       85.5   
Caterpie             1        False   Green       True      0.30        2.9   
Metapod              1        False   Green       True      0.71        9.9   
Butterfree           1        False   White       True      1.09       32.0   
Weedle               1        False   Brown       True      0.30        3.2   
Kakuna               1        False  Yellow       True      0.61       10.0   
Beedrill             1        False  Yellow       True      0.99       29.5   
Pidgey               1        False   Brown       True      0.30        1.8   
Pidgeotto            1        False   Brown       True      1.09       30.0   
Pidgeot              1        False   Brown       True      1.50       39.5   
Rattata              1        False  Purple       True      0.30        3.5   
Raticate             1        False   Brown       True      0.71       18.5   

            Catch_Rate       Body_Style  
Name                                     
Bulbasaur           45        quadruped  
Ivysaur             45        quadruped  
Venusaur            45        quadruped  
Charmander          45   bipedal_tailed  
Charmeleon          45   bipedal_tailed  
Charizard           45   bipedal_tailed  
Squirtle            45   bipedal_tailed  
Wartortle           45   bipedal_tailed  
Blastoise           45   bipedal_tailed  
Caterpie           255        insectoid  
Metapod            120  serpentine_body  
Butterfree          45       four_wings  
Weedle             255  serpentine_body  
Kakuna             120  serpentine_b

<br>
<br>
<font color = '#5a86ad'>**Q. 데이터를 보니, Name보다는 Number가 Index로 더 적합해 보입니다. 최종적으로 index_col이 'Number'가 되도록 pokemon.csv를 불러와보세요!**</font>

<font color = '#77926f'> **Tip -  원본 데이터는 덮어쓰지 마세요. 앞으로 배울 많은 method에 inplace라는 옵션이 있어 기본적으로 덮어쓸 수 없게 설계되긴 했지만, 이름 짓기 귀찮다고 같은 이름에 변형된 데이터를 할당해서는 안됩니다.** </font>

<br>
<br>
<br>

---
#### * DataFrame 저장
변형된 데이터셋의 저장은 <font color="#ff5b00">{저장할 데이터프레임 이름}.to_csv()</font>로 가능합니다. 역시 확장자는 불러올 때와 같이 변경할 수 있습니다.
- 세부 설정
    - 그냥 저장하면 불러올 때 추가로 생성된 index가 같이 저장 됩니다. index = False를 잊지마세요.
    - 그밖의 설정은 shift+tab을 참고하세요.
    


In [17]:
pokemon.to_csv("data/pokemon2.csv")

In [18]:
pkm = pd.read_csv("data/pokemon2.csv")
pkm.head()

Number        Name Type_1  Type_2  Total  HP  Attack  Defense  Speed  \
0       1   Bulbasaur  Grass  Poison    318  45      49       49     45   
1       2     Ivysaur  Grass  Poison    405  60      62       63     60   
2       3    Venusaur  Grass  Poison    525  80      82       83     80   
3       4  Charmander   Fire     NaN    309  39      52       43     65   
4       5  Charmeleon   Fire     NaN    405  58      64       58     80   

   Generation  isLegendary  Color  hasGender  Height_m  Weight_kg  Catch_Rate  \
0           1        False  Green       True      0.71        6.9          45   
1           1        False  Green       True      0.99       13.0          45   
2           1        False  Green       True      2.01      100.0          45   
3           1        False    Red       True      0.61        8.5          45   
4           1        False    Red       True      1.09       19.0          45   

       Body_Style  
0       quadruped  
1       quadruped  
2       quadruped  
3  bipedal_tailed  
4  bipedal_tailed

In [19]:
#새로 생성된 인덱스까지 저장하고 불러오는 경우

pkm.to_csv("data/pokemon3.csv") 

In [20]:
# 인덱스 컬럼이 저잗되고 새 인덱스가 생긴 것을 확인

pkm = pd.read_csv("data/pokemon3.csv")
pkm.head()

Unnamed: 0  Number        Name Type_1  Type_2  Total  HP  Attack  Defense  \
0           0       1   Bulbasaur  Grass  Poison    318  45      49       49   
1           1       2     Ivysaur  Grass  Poison    405  60      62       63   
2           2       3    Venusaur  Grass  Poison    525  80      82       83   
3           3       4  Charmander   Fire     NaN    309  39      52       43   
4           4       5  Charmeleon   Fire     NaN    405  58      64       58   

   Speed  Generation  isLegendary  Color  hasGender  Height_m  Weight_kg  \
0     45           1        False  Green       True      0.71        6.9   
1     60           1        False  Green       True      0.99       13.0   
2     80           1        False  Green       True      2.01      100.0   
3     65           1        False    Red       True      0.61        8.5   
4     80           1        False    Red       True      1.09       19.0   

   Catch_Rate      Body_Style  
0          45       quadruped  
1          45       quadruped  
2          45       quadruped  
3          45  bipedal_tailed  
4          45  bipedal_tailed

---
#### * DataFrame 생성
Nested Array(list)를 DataFrame으로 바꿀 수 있습니다.<br>
pd.Series()로 Array(list)를 Series로 만들 수도 있습니다.<br>

- <font color="#ff5b00">pd.DataFrame()</font>
- 세부 설정
    - columns : columns의 label
    - index : index의 label
- <font color="#ff5b00">pd.Series()</font>
<br>
<br>



In [ ]:
ls = [[1,2],[3,4],[5,6]]

In [ ]:
pd.DataFrame(ls)

In [ ]:
pd.DataFrame(ls,
             columns = ['1열','2열'],
             index = ['1행', '2행', '3행'])

In [ ]:
pd.Series(ls)

---
# 2. Data 살펴보기

---
#### * head, tail

- <font color="#ff5b00">.head()</font>
- <font color="#ff5b00">.tail()</font>

간단히 데이터의 앞뒤를 확인해볼 수 있습니다.<br>
확인을 위해 데이터 전체를 불러올 경우 컬럼과 그 대표값을 대략적으로 확인하기 어렵기 때문에 자주 쓰입니다.<br>

In [21]:
pokemon = pd.read_csv("data/pokemon.csv", index_col = "Number")
pokemon.head()

Name Type_1  Type_2  Total  HP  Attack  Defense  Speed  \
Number                                                                 
1        Bulbasaur  Grass  Poison    318  45      49       49     45   
2          Ivysaur  Grass  Poison    405  60      62       63     60   
3         Venusaur  Grass  Poison    525  80      82       83     80   
4       Charmander   Fire     NaN    309  39      52       43     65   
5       Charmeleon   Fire     NaN    405  58      64       58     80   

        Generation  isLegendary  Color  hasGender  Height_m  Weight_kg  \
Number                                                                   
1                1        False  Green       True      0.71        6.9   
2                1        False  Green       True      0.99       13.0   
3                1        False  Green       True      2.01      100.0   
4                1        False    Red       True      0.61        8.5   
5                1        False    Red       True      1.09       19.0   

        Catch_Rate      Body_Style  
Number                              
1               45       quadruped  
2               45       quadruped  
3               45       quadruped  
4               45  bipedal_tailed  
5               45  bipedal_tailed

확인할 갯수도 설정할 수 있습니다. 기본은 5개인데, 좀 더 많이 볼 수 있을까요?

In [22]:
pokemon.head(20)

Name  Type_1  Type_2  Total  HP  Attack  Defense  Speed  \
Number                                                                  
1        Bulbasaur   Grass  Poison    318  45      49       49     45   
2          Ivysaur   Grass  Poison    405  60      62       63     60   
3         Venusaur   Grass  Poison    525  80      82       83     80   
4       Charmander    Fire     NaN    309  39      52       43     65   
5       Charmeleon    Fire     NaN    405  58      64       58     80   
6        Charizard    Fire  Flying    534  78      84       78    100   
7         Squirtle   Water     NaN    314  44      48       65     43   
8        Wartortle   Water     NaN    405  59      63       80     58   
9        Blastoise   Water     NaN    530  79      83      100     78   
10        Caterpie     Bug     NaN    195  45      30       35     45   
11         Metapod     Bug     NaN    205  50      20       55     30   
12      Butterfree     Bug  Flying    395  60      45       50     70   
13          Weedle     Bug  Poison    195  40      35       30     50   
14          Kakuna     Bug  Poison    205  45      25       50     35   
15        Beedrill     Bug  Poison    395  65      90       40     75   
16          Pidgey  Normal  Flying    251  40      45       40     56   
17       Pidgeotto  Normal  Flying    349  63      60       55     71   
18         Pidgeot  Normal  Flying    479  83      80       75    101   
19         Rattata  Normal     NaN    253  30      56       35     72   
20        Raticate  Normal     NaN    413  55      81       60     97   

        Generation  isLegendary   Color  hasGender  Height_m  Weight_kg  \
Number                                                                    
1                1        False   Green       True      0.71        6.9   
2                1        False   Green       True      0.99       13.0   
3                1        False   Green       True      2.01      100.0   
4                1        False     Red       True      0.61        8.5   
5                1        False     Red       True      1.09       19.0   
6                1        False     Red       True      1.70       90.5   
7                1        False    Blue       True      0.51        9.0   
8                1        False    Blue       True      0.99       22.5   
9                1        False    Blue       True      1.60       85.5   
10               1        False   Green       True      0.30        2.9   
11               1        False   Green       True      0.71        9.9   
12               1        False   White       True      1.09       32.0   
13               1        False   Brown       True      0.30        3.2   
14               1        False  Yellow       True      0.61       10.0   
15               1        False  Yellow       True      0.99       29.5   
16               1        False   Brown       True      0.30        1.8   
17               1        False   Brown       True      1.09       30.0   
18               1        False   Brown       True      1.50       39.5   
19               1        False  Purple       True      0.30        3.5   
20               1        False   Brown       True      0.71       18.5   

        Catch_Rate       Body_Style  
Number                               
1               45        quadruped  
2               45        quadruped  
3               45        quadruped  
4               45   bipedal_tailed  
5               45   bipedal_tailed  
6               45   bipedal_tailed  
7               45   bipedal_tailed  
8               45   bipedal_tailed  
9               45   bipedal_tailed  
10             255        insectoid  
11             120  serpentine_body  
12              45       four_wings  
13             255  serpentine_body  
14             120  serpentine_body  
15              45       four_wings  
16             255        two_wings  
17             120        two_wings  
18              45        two_wing

In [23]:
pokemon.tail()

Name   Type_1  Type_2  Total   HP  Attack  Defense  Speed  \
Number                                                                   
717       Yveltal     Dark  Flying    680  126     131       95     99   
718       Zygarde   Dragon  Ground    600  108     100      121     95   
719       Diancie     Rock   Fairy    600   50     100      150     50   
720         Hoopa  Psychic   Ghost    600   80     110       60     70   
721     Volcanion     Fire   Water    600   80     110      120     70   

        Generation  isLegendary   Color  hasGender  Height_m  Weight_kg  \
Number                                                                    
717              6         True     Red      False      5.79      203.0   
718              6         True   Green      False      5.00      305.0   
719              6         True    Pink      False      0.71        8.8   
720              6         True  Purple      False      0.51        9.0   
721              6         True   Brown      False      1.70      195.0   

        Catch_Rate       Body_Style  
Number                               
717             45        two_wings  
718              3  serpentine_body  
719              3        head_arms  
720              3        head_only  
721              3        quadruped

---
#### * 한 column만 보기

- <font color="#ff5b00">df["column_name"]</font>
- <font color="#ff5b00">df.column_name</font>

데이터의 특정 column을 retrieve 합니다.

In [24]:
pokemon["Name"]
# pokemon.Name

Number
1       Bulbasaur
2         Ivysaur
3        Venusaur
4      Charmander
5      Charmeleon
6       Charizard
7        Squirtle
8       Wartortle
9       Blastoise
10       Caterpie
11        Metapod
12     Butterfree
13         Weedle
14         Kakuna
15       Beedrill
16         Pidgey
17      Pidgeotto
18        Pidgeot
19        Rattata
20       Raticate
21        Spearow
22         Fearow
23          Ekans
24          Arbok
25        Pikachu
26         Raichu
27      Sandshrew
28      Sandslash
29       Nidoran♀
30       Nidorina
          ...    
692     Clauncher
693     Clawitzer
694    Helioptile
695     Heliolisk
696        Tyrunt
697     Tyrantrum
698        Amaura
699       Aurorus
700       Sylveon
701      Hawlucha
702       Dedenne
703       Carbink
704         Goomy
705       Sliggoo
706        Goodra
707        Klefki
708      Phantump
709     Trevenant
710     Pumpkaboo
711     Gourgeist
712      Bergmite
713       Avalugg
714        Noibat
715       Noivern
716

---
<br>
<br>

#### * column, index보기
- <font color="#ff5b00">.columns()</font>
- <font color="#ff5b00">.index()</font>

DataFrame의 columns와 index를 확인할 수 있습니다.

In [25]:
pokemon.columns

Index(['Name', 'Type_1', 'Type_2', 'Total', 'HP', 'Attack', 'Defense', 'Speed',
       'Generation', 'isLegendary', 'Color', 'hasGender', 'Height_m',
       'Weight_kg', 'Catch_Rate', 'Body_Style'],
      dtype='object')

In [26]:
pokemon.index

Int64Index([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,
            ...
            712, 713, 714, 715, 716, 717, 718, 719, 720, 721],
           dtype='int64', name='Number', length=721)

<br>
<br>
index의 이름이 마음에 들지 않을 경우 수정하거나 지울 수 있습니다.

In [27]:
pokemon.index.name = 'Pokemon_Number'

In [28]:
pokemon.head()

Name Type_1  Type_2  Total  HP  Attack  Defense  Speed  \
Pokemon_Number                                                                 
1                Bulbasaur  Grass  Poison    318  45      49       49     45   
2                  Ivysaur  Grass  Poison    405  60      62       63     60   
3                 Venusaur  Grass  Poison    525  80      82       83     80   
4               Charmander   Fire     NaN    309  39      52       43     65   
5               Charmeleon   Fire     NaN    405  58      64       58     80   

                Generation  isLegendary  Color  hasGender  Height_m  \
Pokemon_Number                                                        
1                        1        False  Green       True      0.71   
2                        1        False  Green       True      0.99   
3                        1        False  Green       True      2.01   
4                        1        False    Red       True      0.61   
5                        1        False    Red       True      1.09   

                Weight_kg  Catch_Rate      Body_Style  
Pokemon_Number                                         
1                     6.9          45       quadruped  
2                    13.0          45       quadruped  
3                   100.0          45       quadruped  
4                     8.5          45  bipedal_tailed  
5                    19.0          45  bipedal_tailed

In [29]:
del pokemon.index.name

In [30]:
pokemon.head()

Name Type_1  Type_2  Total  HP  Attack  Defense  Speed  Generation  \
1   Bulbasaur  Grass  Poison    318  45      49       49     45           1   
2     Ivysaur  Grass  Poison    405  60      62       63     60           1   
3    Venusaur  Grass  Poison    525  80      82       83     80           1   
4  Charmander   Fire     NaN    309  39      52       43     65           1   
5  Charmeleon   Fire     NaN    405  58      64       58     80           1   

   isLegendary  Color  hasGender  Height_m  Weight_kg  Catch_Rate  \
1        False  Green       True      0.71        6.9          45   
2        False  Green       True      0.99       13.0          45   
3        False  Green       True      2.01      100.0          45   
4        False    Red       True      0.61        8.5          45   
5        False    Red       True      1.09       19.0          45   

       Body_Style  
1       quadruped  
2       quadruped  
3       quadruped  
4  bipedal_tailed  
5  bipedal_tailed

<br>
<br>
Row나 Column의 개수가 너무 많아서 모든 Row, Column이 보이지 않을 때도 있습니다.
그럴 때에는 출력할 row와 column의 개수를 지정할 수 있습니다.

- <font color="#ff5b00">pd.options.display.max_columns</font>
- <font color="#ff5b00">pd.set_option('max_columns', 999)</font>

In [31]:
#row의 숫자를 바꾸고 싶다면 columns대신 rows라고 하면 됩니다!
pd.options.display.max_columns = 999

In [32]:
# pd.set_option에서 다양한 설정을 변경할 수 있습니다.

pd.set_option('max_columns', 999)
pd.set_option('max_colwidth', 800)

---
<br>
#### * DataFrame 정보 확인 
- <font color="#ff5b00">.info()</font>
- <font color="#ff5b00">.isnull()</font>


DataFrame에 담긴 정보를 요약하여 볼 수 있습니다.<br>
보통 각 column에 담긴 데이터의 타입(범주형 or 연속형), 결측치 등을 확인합니다.<br>

In [33]:
#데이터 타입 확인

pokemon.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 721 entries, 1 to 721
Data columns (total 16 columns):
Name           721 non-null object
Type_1         721 non-null object
Type_2         350 non-null object
Total          721 non-null int64
HP             721 non-null int64
Attack         721 non-null int64
Defense        721 non-null int64
Speed          721 non-null int64
Generation     721 non-null int64
isLegendary    721 non-null bool
Color          721 non-null object
hasGender      721 non-null bool
Height_m       721 non-null float64
Weight_kg      721 non-null float64
Catch_Rate     721 non-null int64
Body_Style     721 non-null object
dtypes: bool(2), float64(2), int64(7), object(5)
memory usage: 85.9+ KB


In [34]:
# 결측치 확인, Series return

# pokemon.isnull().any()
pokemon.isnull().sum()

Name             0
Type_1           0
Type_2         371
Total            0
HP               0
Attack           0
Defense          0
Speed            0
Generation       0
isLegendary      0
Color            0
hasGender        0
Height_m         0
Weight_kg        0
Catch_Rate       0
Body_Style       0
dtype: int64

<br>
<br>
<font color = '#5a86ad'>**Q. 결측치의 절대적 수치도 중요하지만, 전체 데이터에 비해 상대적으로 결측치가 어느 정도 되는지를 파악하는 것도 중요합니다. <br>전체 데이터 수 대비 결측치의 %를 보려면 어떻게 해야 할까요?**</font>

Tip: array의 길이는 len(array)로 찾을 수 있습니다.
<br>
Tip: python의 사칙연산은 + - * /

<br>

이처럼 확인된 결측치들은 어떻게 처리하면 좋을까요?<br>
데이터나 적용할 모델에 따라 다르겠으나, 그 수가 많지 않을 경우 그냥 0으로 채워주거나 결측치가 들어있는 row를 지워버립니다.<br>
특정 변수에만 결측치가 많다면 해당 column만을 지울 수도 있을 것입니다.<br>
또 다른 방법으로는 평균값,해당 데이터와 인접한 데이터의 값을 넣거나(KNN), 특정 모델로 예측한 값을 넣는 것도 생각해볼 수 있겠죠.<br>
일단 포켓몬 데이터의 결측치는 범주형 데이터이고 값이 없는게 정상이기 때문에 0으로 채우기, 행/열 지우기 정도만 연습해보겠습니다.


- <font color="#ff5b00">.dropna()</font>
- <font color="#ff5b00">.fillna()</font>

In [36]:
# thresh -> 결측치가 n개 이상인 경우에만 drop

print(
    len(pokemon.dropna()),
    len(pokemon.dropna(thresh=2))) #살아남은 수를 보여줘요!

350 721


In [37]:
# axis=1인 경우 열 기준으로 drop -> Type_2가 사라진 것을 확인 (하나라도 없는 열 drop)

pokemon.dropna(axis = 1).head()

Name Type_1  Total  HP  Attack  Defense  Speed  Generation  \
1   Bulbasaur  Grass    318  45      49       49     45           1   
2     Ivysaur  Grass    405  60      62       63     60           1   
3    Venusaur  Grass    525  80      82       83     80           1   
4  Charmander   Fire    309  39      52       43     65           1   
5  Charmeleon   Fire    405  58      64       58     80           1   

   isLegendary  Color  hasGender  Height_m  Weight_kg  Catch_Rate  \
1        False  Green       True      0.71        6.9          45   
2        False  Green       True      0.99       13.0          45   
3        False  Green       True      2.01      100.0          45   
4        False    Red       True      0.61        8.5          45   
5        False    Red       True      1.09       19.0          45   

       Body_Style  
1       quadruped  
2       quadruped  
3       quadruped  
4  bipedal_tailed  
5  bipedal_tailed

In [38]:
print(
    len(pokemon.dropna(axis = 1).columns),
    len(pokemon.columns))

15 16


In [39]:
# 결측치를 0으로 채우기

pokemon.fillna(0).head()

Name Type_1  Type_2  Total  HP  Attack  Defense  Speed  Generation  \
1   Bulbasaur  Grass  Poison    318  45      49       49     45           1   
2     Ivysaur  Grass  Poison    405  60      62       63     60           1   
3    Venusaur  Grass  Poison    525  80      82       83     80           1   
4  Charmander   Fire       0    309  39      52       43     65           1   
5  Charmeleon   Fire       0    405  58      64       58     80           1   

   isLegendary  Color  hasGender  Height_m  Weight_kg  Catch_Rate  \
1        False  Green       True      0.71        6.9          45   
2        False  Green       True      0.99       13.0          45   
3        False  Green       True      2.01      100.0          45   
4        False    Red       True      0.61        8.5          45   
5        False    Red       True      1.09       19.0          45   

       Body_Style  
1       quadruped  
2       quadruped  
3       quadruped  
4  bipedal_tailed  
5  bipedal_tailed

In [40]:
# 0이 아닌 다른 값으로도 채울 수 있음

pokemon.fillna("None").head()

Name Type_1  Type_2  Total  HP  Attack  Defense  Speed  Generation  \
1   Bulbasaur  Grass  Poison    318  45      49       49     45           1   
2     Ivysaur  Grass  Poison    405  60      62       63     60           1   
3    Venusaur  Grass  Poison    525  80      82       83     80           1   
4  Charmander   Fire    None    309  39      52       43     65           1   
5  Charmeleon   Fire    None    405  58      64       58     80           1   

   isLegendary  Color  hasGender  Height_m  Weight_kg  Catch_Rate  \
1        False  Green       True      0.71        6.9          45   
2        False  Green       True      0.99       13.0          45   
3        False  Green       True      2.01      100.0          45   
4        False    Red       True      0.61        8.5          45   
5        False    Red       True      1.09       19.0          45   

       Body_Style  
1       quadruped  
2       quadruped  
3       quadruped  
4  bipedal_tailed  
5  bipedal_tailed

<br>

결측치 확인 후 간단한 통계량과 상관관계, 빈도 등을 확인하는 방법입니다.<br>
꽤 자주 쓰이니 꼭 알아두세요!<br>

- <font color="#ff5b00">.describe()</font>
- <font color="#ff5b00">.corr()</font>
- <font color="#ff5b00">{Series}.value_counts()</font>
- <font color="#ff5b00">{Series}.unique()</font>

<br>

In [41]:
# 각 컬럼 별로 간단한 통계값 제공

pokemon.describe()

Total          HP      Attack     Defense       Speed  Generation  \
count  721.000000  721.000000  721.000000  721.000000  721.000000  721.000000   
mean   417.945908   68.380028   75.013870   70.808599   65.714286    3.323162   
std    109.663671   25.848272   28.984475   29.296558   27.277920    1.669873   
min    180.000000    1.000000    5.000000    5.000000    5.000000    1.000000   
25%    320.000000   50.000000   53.000000   50.000000   45.000000    2.000000   
50%    424.000000   65.000000   74.000000   65.000000   65.000000    3.000000   
75%    499.000000   80.000000   95.000000   85.000000   85.000000    5.000000   
max    720.000000  255.000000  165.000000  230.000000  160.000000    6.000000   

         Height_m   Weight_kg  Catch_Rate  
count  721.000000  721.000000  721.000000  
mean     1.144979   56.773370  100.246879  
std      1.044369   89.095667   76.573513  
min      0.100000    0.100000    3.000000  
25%      0.610000    9.400000   45.000000  
50%      0.990000   28.000000   65.000000  
75%      1.400000   61.000000  180.000000  
max     14.500000  950.000000  255.000000

In [42]:
# 각 변수간 상관관계

pokemon.corr()

Total        HP    Attack   Defense     Speed  Generation  \
Total        1.000000  0.642628  0.704164  0.605831  0.548890    0.092868   
HP           0.642628  1.000000  0.431680  0.228834  0.170031    0.071545   
Attack       0.704164  0.431680  1.000000  0.433233  0.335013    0.093857   
Defense      0.605831  0.228834  0.433233  1.000000 -0.008663    0.068409   
Speed        0.548890  0.170031  0.335013 -0.008663  1.000000    0.003920   
Generation   0.092868  0.071545  0.093857  0.068409  0.003920    1.000000   
isLegendary  0.481837  0.258926  0.302786  0.274446  0.286082    0.071875   
hasGender   -0.385978 -0.155031 -0.196892 -0.269466 -0.216964   -0.029916   
Height_m     0.526813  0.442872  0.408590  0.354205  0.224617   -0.051304   
Weight_kg    0.535966  0.431320  0.469395  0.476983  0.108637    0.034003   
Catch_Rate  -0.738280 -0.478725 -0.525106 -0.436558 -0.410557   -0.025227   

             isLegendary  hasGender  Height_m  Weight_kg  Catch_Rate  
Total           0.481837  -0.385978  0.526813   0.535966   -0.738280  
HP              0.258926  -0.155031  0.442872   0.431320   -0.478725  
Attack          0.302786  -0.196892  0.408590   0.469395   -0.525106  
Defense         0.274446  -0.269466  0.354205   0.476983   -0.436558  
Speed           0.286082  -0.216964  0.224617   0.108637   -0.410557  
Generation      0.071875  -0.029916 -0.051304   0.034003   -0.025227  
isLegendary     1.000000  -0.644714  0.326323   0.425219   -0.319302  
hasGender      -0.644714   1.000000 -0.200026  -0.361465    0.272304  
Height_m        0.326323  -0.200026  1.000000   0.661342   -0.382862  
Weight_kg       0.425219  -0.361465  0.661342   1.000000   -0.367798  
Catch_Rate     -0.319302   0.272304 -0.382862  -0.367798    1.000000

In [43]:
#각 Series에서 어떤 값들이 몇번 나타났는지 확인할 수 있다.

pokemon['Type_1'].value_counts()

Water       105
Normal       93
Grass        66
Bug          63
Fire         47
Psychic      47
Rock         41
Electric     36
Ground       30
Poison       28
Dark         28
Fighting     25
Dragon       24
Ice          23
Ghost        23
Steel        22
Fairy        17
Flying        3
Name: Type_1, dtype: int64

In [44]:
#중복을 제거한 value들을 보고싶을 때

pokemon['Type_2'].unique()

array(['Poison', nan, 'Flying', 'Ground', 'Fairy', 'Grass', 'Fighting',
       'Psychic', 'Steel', 'Ice', 'Rock', 'Water', 'Electric', 'Fire',
       'Dragon', 'Dark', 'Ghost', 'Bug', 'Normal'], dtype=object)

---

# 3. Columns 생성, 연산, 이름 수정, 삭제
---

#### * column 생성(1) - 단순 대입
- <font color="#ff5b00">{DF}['new column name'] = [Array나 list]</font>
- <font color="#ff5b00">{DF}.assign( new column name = [Array나 list] )</font>



In [45]:
pokemon.assign( Random_status = np.random.randint(low=1,high=100, size=721)).head()

Name Type_1  Type_2  Total  HP  Attack  Defense  Speed  Generation  \
1   Bulbasaur  Grass  Poison    318  45      49       49     45           1   
2     Ivysaur  Grass  Poison    405  60      62       63     60           1   
3    Venusaur  Grass  Poison    525  80      82       83     80           1   
4  Charmander   Fire     NaN    309  39      52       43     65           1   
5  Charmeleon   Fire     NaN    405  58      64       58     80           1   

   isLegendary  Color  hasGender  Height_m  Weight_kg  Catch_Rate  \
1        False  Green       True      0.71        6.9          45   
2        False  Green       True      0.99       13.0          45   
3        False  Green       True      2.01      100.0          45   
4        False    Red       True      0.61        8.5          45   
5        False    Red       True      1.09       19.0          45   

       Body_Style  Random_status  
1       quadruped             12  
2       quadruped             37  
3       quadruped             67  
4  bipedal_tailed             30  
5  bipedal_tailed             26

In [46]:
pokemon.head()

Name Type_1  Type_2  Total  HP  Attack  Defense  Speed  Generation  \
1   Bulbasaur  Grass  Poison    318  45      49       49     45           1   
2     Ivysaur  Grass  Poison    405  60      62       63     60           1   
3    Venusaur  Grass  Poison    525  80      82       83     80           1   
4  Charmander   Fire     NaN    309  39      52       43     65           1   
5  Charmeleon   Fire     NaN    405  58      64       58     80           1   

   isLegendary  Color  hasGender  Height_m  Weight_kg  Catch_Rate  \
1        False  Green       True      0.71        6.9          45   
2        False  Green       True      0.99       13.0          45   
3        False  Green       True      2.01      100.0          45   
4        False    Red       True      0.61        8.5          45   
5        False    Red       True      1.09       19.0          45   

       Body_Style  
1       quadruped  
2       quadruped  
3       quadruped  
4  bipedal_tailed  
5  bipedal_tailed

In [47]:
pokemon['Random_status'] = np.random.randint(low=1,high=100, size=721)

In [48]:
# DF 맨 뒤에 새로운 변수가 생성된 것을 확인

pokemon.head()

Name Type_1  Type_2  Total  HP  Attack  Defense  Speed  Generation  \
1   Bulbasaur  Grass  Poison    318  45      49       49     45           1   
2     Ivysaur  Grass  Poison    405  60      62       63     60           1   
3    Venusaur  Grass  Poison    525  80      82       83     80           1   
4  Charmander   Fire     NaN    309  39      52       43     65           1   
5  Charmeleon   Fire     NaN    405  58      64       58     80           1   

   isLegendary  Color  hasGender  Height_m  Weight_kg  Catch_Rate  \
1        False  Green       True      0.71        6.9          45   
2        False  Green       True      0.99       13.0          45   
3        False  Green       True      2.01      100.0          45   
4        False    Red       True      0.61        8.5          45   
5        False    Red       True      1.09       19.0          45   

       Body_Style  Random_status  
1       quadruped             95  
2       quadruped             33  
3       quadruped             93  
4  bipedal_tailed             26  
5  bipedal_tailed             39

---
#### * column 생성(2) - 사칙연산도 가능합니다

판다스의 재밌는 점은 column간 사칙연산이 가능하다는 것입니다.<br>
간단한 연산으로 새로운 변수를 생성하거나 기존 변수를 바꿀 수 있습니다.<br>



In [49]:
pokemon['Att+Def'] = 0.4*pokemon['Attack'] + 0.6*pokemon['Defense']
pokemon.head()

Name Type_1  Type_2  Total  HP  Attack  Defense  Speed  Generation  \
1   Bulbasaur  Grass  Poison    318  45      49       49     45           1   
2     Ivysaur  Grass  Poison    405  60      62       63     60           1   
3    Venusaur  Grass  Poison    525  80      82       83     80           1   
4  Charmander   Fire     NaN    309  39      52       43     65           1   
5  Charmeleon   Fire     NaN    405  58      64       58     80           1   

   isLegendary  Color  hasGender  Height_m  Weight_kg  Catch_Rate  \
1        False  Green       True      0.71        6.9          45   
2        False  Green       True      0.99       13.0          45   
3        False  Green       True      2.01      100.0          45   
4        False    Red       True      0.61        8.5          45   
5        False    Red       True      1.09       19.0          45   

       Body_Style  Random_status  Att+Def  
1       quadruped             95     49.0  
2       quadruped             33     62.6  
3       quadruped             93     82.6  
4  bipedal_tailed             26     46.6  
5  bipedal_tailed             39     60.4

<br>

새로 만들고 싶은 column이 사칙연산으로 만들어질 수 없는 경우?<br>
lambda와 map을 잘 이용하면 쉽습니다.<br>
replace는 특정 값을 그에 대응되는 다른 값으로 변환할 때 사용합니다.<br>

- <font color="#ff5b00">.map()</font>
- <font color="#ff5b00">.replace()</font>
- <font color="#ff5b00">lambda</font>

<br>

In [50]:
# 예를 들면...
# 공격력 600 이상의 강한 포켓몬만 따로 표시해두고 싶을 때...

pokemon['Strong?'] = pokemon['Total'].map(lambda x: 'Y' if x>600 else 'N') #lambda x: .map()이 불러진 대상을 x라고 칭하는 것! 

In [51]:
pokemon.head()

Name Type_1  Type_2  Total  HP  Attack  Defense  Speed  Generation  \
1   Bulbasaur  Grass  Poison    318  45      49       49     45           1   
2     Ivysaur  Grass  Poison    405  60      62       63     60           1   
3    Venusaur  Grass  Poison    525  80      82       83     80           1   
4  Charmander   Fire     NaN    309  39      52       43     65           1   
5  Charmeleon   Fire     NaN    405  58      64       58     80           1   

   isLegendary  Color  hasGender  Height_m  Weight_kg  Catch_Rate  \
1        False  Green       True      0.71        6.9          45   
2        False  Green       True      0.99       13.0          45   
3        False  Green       True      2.01      100.0          45   
4        False    Red       True      0.61        8.5          45   
5        False    Red       True      1.09       19.0          45   

       Body_Style  Random_status  Att+Def Strong?  
1       quadruped             95     49.0       N  
2       quadruped             33     62.6       N  
3       quadruped             93     82.6       N  
4  bipedal_tailed             26     46.6       N  
5  bipedal_tailed             39     60.4       N

In [52]:
# replace 활용 예시

pokemon.Generation.replace(
    [1, 2, 3, 4, 5, 6], 
    ['first','second','third','fourth','fifth','sixth']).head()

1    first
2    first
3    first
4    first
5    first
Name: Generation, dtype: object

---
#### * column 생성(3) - column 이름 수정

- <font color="#ff5b00">.rename(columns = {'Name' : 'name'}) ##기존이름 : 새로운이름</font>
- <font color="#ff5b00">.columns = [new columns][</font>



In [53]:
# 원하는 컬럼 이름이 바뀐 데이터프레임을 반환

pokemon.rename(columns = {'Attack' : 'ATK'}).head()

Name Type_1  Type_2  Total  HP  ATK  Defense  Speed  Generation  \
1   Bulbasaur  Grass  Poison    318  45   49       49     45           1   
2     Ivysaur  Grass  Poison    405  60   62       63     60           1   
3    Venusaur  Grass  Poison    525  80   82       83     80           1   
4  Charmander   Fire     NaN    309  39   52       43     65           1   
5  Charmeleon   Fire     NaN    405  58   64       58     80           1   

   isLegendary  Color  hasGender  Height_m  Weight_kg  Catch_Rate  \
1        False  Green       True      0.71        6.9          45   
2        False  Green       True      0.99       13.0          45   
3        False  Green       True      2.01      100.0          45   
4        False    Red       True      0.61        8.5          45   
5        False    Red       True      1.09       19.0          45   

       Body_Style  Random_status  Att+Def Strong?  
1       quadruped             95     49.0       N  
2       quadruped             33     62.6       N  
3       quadruped             93     82.6       N  
4  bipedal_tailed             26     46.6       N  
5  bipedal_tailed             39     60.4       N

In [54]:
pokemon.head()

Name Type_1  Type_2  Total  HP  Attack  Defense  Speed  Generation  \
1   Bulbasaur  Grass  Poison    318  45      49       49     45           1   
2     Ivysaur  Grass  Poison    405  60      62       63     60           1   
3    Venusaur  Grass  Poison    525  80      82       83     80           1   
4  Charmander   Fire     NaN    309  39      52       43     65           1   
5  Charmeleon   Fire     NaN    405  58      64       58     80           1   

   isLegendary  Color  hasGender  Height_m  Weight_kg  Catch_Rate  \
1        False  Green       True      0.71        6.9          45   
2        False  Green       True      0.99       13.0          45   
3        False  Green       True      2.01      100.0          45   
4        False    Red       True      0.61        8.5          45   
5        False    Red       True      1.09       19.0          45   

       Body_Style  Random_status  Att+Def Strong?  
1       quadruped             95     49.0       N  
2       quadruped             33     62.6       N  
3       quadruped             93     82.6       N  
4  bipedal_tailed             26     46.6       N  
5  bipedal_tailed             39     60.4       N

In [55]:
# 원본의 컬럼 이름을 직접 바꿉니다.
# 이게 rename보다 더 편하긴 하지만, inplace되기 때문에 추천하지 않습니다.

pokemon.columns = ['Name', 'Type_1', 'Type_2', 'Total', 'HP', 'ATK', 'Defense', 'Speed',
       'Generation', 'isLegendary', 'Color', 'hasGender', 'Height_m',
       'Weight_kg', 'Catch_Rate', 'Body_Style', 'Random_status', 'Att+Def','Strong?']

In [56]:
pokemon.head()

Name Type_1  Type_2  Total  HP  ATK  Defense  Speed  Generation  \
1   Bulbasaur  Grass  Poison    318  45   49       49     45           1   
2     Ivysaur  Grass  Poison    405  60   62       63     60           1   
3    Venusaur  Grass  Poison    525  80   82       83     80           1   
4  Charmander   Fire     NaN    309  39   52       43     65           1   
5  Charmeleon   Fire     NaN    405  58   64       58     80           1   

   isLegendary  Color  hasGender  Height_m  Weight_kg  Catch_Rate  \
1        False  Green       True      0.71        6.9          45   
2        False  Green       True      0.99       13.0          45   
3        False  Green       True      2.01      100.0          45   
4        False    Red       True      0.61        8.5          45   
5        False    Red       True      1.09       19.0          45   

       Body_Style  Random_status  Att+Def Strong?  
1       quadruped             95     49.0       N  
2       quadruped             33     62.6       N  
3       quadruped             93     82.6       N  
4  bipedal_tailed             26     46.6       N  
5  bipedal_tailed             39     60.4       N

---
# 4. Indexing, Slicing
---

#### * Indexing : list의 indexing과 마찬가지로, 내가 필요한 부분만 편집하여 보는 것을 뜻합니다.
단순히 특정 지점부터 n번째까지만 보는 것은 물론, **특정 조건**에 따라 indexing하는 것도 가능합니다.

- <font color="#ff5b00">{DF}[['colname1','colname2', ... ]] # 특정 column만 보기, series에서의 인덱싱과 다름</font>
- <font color="#ff5b00">{DF}[int:int] # row 기준으로 인덱싱</font>



In [57]:
pokemon[['Name','Total']].head()

Name  Total
1   Bulbasaur    318
2     Ivysaur    405
3    Venusaur    525
4  Charmander    309
5  Charmeleon    405

---
#### * DataFrame에서 2번째 열부터 10번째 열까지만 불러와볼까요?

In [58]:
pokemon[1:10]

Name Type_1  Type_2  Total  HP  ATK  Defense  Speed  Generation  \
2      Ivysaur  Grass  Poison    405  60   62       63     60           1   
3     Venusaur  Grass  Poison    525  80   82       83     80           1   
4   Charmander   Fire     NaN    309  39   52       43     65           1   
5   Charmeleon   Fire     NaN    405  58   64       58     80           1   
6    Charizard   Fire  Flying    534  78   84       78    100           1   
7     Squirtle  Water     NaN    314  44   48       65     43           1   
8    Wartortle  Water     NaN    405  59   63       80     58           1   
9    Blastoise  Water     NaN    530  79   83      100     78           1   
10    Caterpie    Bug     NaN    195  45   30       35     45           1   

    isLegendary  Color  hasGender  Height_m  Weight_kg  Catch_Rate  \
2         False  Green       True      0.99       13.0          45   
3         False  Green       True      2.01      100.0          45   
4         False    Red       True      0.61        8.5          45   
5         False    Red       True      1.09       19.0          45   
6         False    Red       True      1.70       90.5          45   
7         False   Blue       True      0.51        9.0          45   
8         False   Blue       True      0.99       22.5          45   
9         False   Blue       True      1.60       85.5          45   
10        False  Green       True      0.30        2.9         255   

        Body_Style  Random_status  Att+Def Strong?  
2        quadruped             33     62.6       N  
3        quadruped             93     82.6       N  
4   bipedal_tailed             26     46.6       N  
5   bipedal_tailed             39     60.4       N  
6   bipedal_tailed             77     80.4       N  
7   bipedal_tailed             11     58.2       N  
8   bipedal_tailed             22     73.2       N  
9   bipedal_tailed             81     93.2       N  
10       insectoid             95     33.0       N

---
#### * 행의 순서 X 열 순서에 따라 indexing (좌표로 찾아가기)

- <font color="#ff5b00">pokemon.iloc[row number, col number]</font>

In [59]:
pokemon.head(3)

Name Type_1  Type_2  Total  HP  ATK  Defense  Speed  Generation  \
1  Bulbasaur  Grass  Poison    318  45   49       49     45           1   
2    Ivysaur  Grass  Poison    405  60   62       63     60           1   
3   Venusaur  Grass  Poison    525  80   82       83     80           1   

   isLegendary  Color  hasGender  Height_m  Weight_kg  Catch_Rate Body_Style  \
1        False  Green       True      0.71        6.9          45  quadruped   
2        False  Green       True      0.99       13.0          45  quadruped   
3        False  Green       True      2.01      100.0          45  quadruped   

   Random_status  Att+Def Strong?  
1             95     49.0       N  
2             33     62.6       N  
3             93     82.6       N

pokemon의 3번째 행, 5번째 값인 82를 뽑고 싶다면?

In [60]:
pokemon.iloc[2,5]

82

특정 위치의 값을 수정하고 싶다면?

In [61]:
pokemon.iloc[2,5] = 60

In [62]:
pokemon.head(3)

Name Type_1  Type_2  Total  HP  ATK  Defense  Speed  Generation  \
1  Bulbasaur  Grass  Poison    318  45   49       49     45           1   
2    Ivysaur  Grass  Poison    405  60   62       63     60           1   
3   Venusaur  Grass  Poison    525  80   60       83     80           1   

   isLegendary  Color  hasGender  Height_m  Weight_kg  Catch_Rate Body_Style  \
1        False  Green       True      0.71        6.9          45  quadruped   
2        False  Green       True      0.99       13.0          45  quadruped   
3        False  Green       True      2.01      100.0          45  quadruped   

   Random_status  Att+Def Strong?  
1             95     49.0       N  
2             33     62.6       N  
3             93     82.6       N

---
#### * 행의 이름 X 열의 이름에 따라 indexing

- <font color="#ff5b00">pokemon.loc[index, columns]</font>



pokemon에서 1,2 행의 Name과 Total만 보고 싶다면?

In [63]:
pokemon.loc[[1,2],['Name','Total']]

Name  Total
1  Bulbasaur    318
2    Ivysaur    405

---
#### * 조건에 맞는 데이터만 골라보기
DF에 조건문을 입력하여 내가 원하는 조건 하의 데이터만을 가려낼 수 있습니다.<br>

- <font color="#ff5b00">df[내가 원하는 조건문을 넣으세요!] : True인 것만 필터링해줍니다.]</font>
- <font color="#ff5b00">df[df[column] > 2]</font>
- <font color="#ff5b00">df[lambda x: x[column] > 2]</font>

In [64]:
pokemon[pokemon['Total'] > 600].head()

Name   Type_1  Type_2  Total   HP  ATK  Defense  Speed  Generation  \
150   Mewtwo  Psychic     NaN    680  106  110       90    130           1   
249    Lugia  Psychic  Flying    680  106   90      130    110           2   
250    Ho-Oh     Fire  Flying    680  106  130       90     90           2   
289  Slaking   Normal     NaN    670  150  160      100    100           3   
382   Kyogre    Water     NaN    670  100  100       90     90           3   

     isLegendary   Color  hasGender  Height_m  Weight_kg  Catch_Rate  \
150         True  Purple      False      2.01      122.0           3   
249         True   White      False      5.21      216.0           3   
250         True     Red      False      3.81      199.0           3   
289        False   Brown       True      2.01      130.5          45   
382         True    Blue      False      4.50      352.0           3   

           Body_Style  Random_status  Att+Def Strong?  
150    bipedal_tailed             79     98.0       Y  
249         two_wings             16    114.0       Y  
250         two_wings             23    106.0       Y  
289  bipedal_tailless             69    124.0       Y  
382         with_fins             38     94.0       Y

lambda를 활용하면 DF의 이름을 반복하지 않아도 됩니다.

In [65]:
pokemon[lambda x : x["Total"] > 600].head()

Name   Type_1  Type_2  Total   HP  ATK  Defense  Speed  Generation  \
150   Mewtwo  Psychic     NaN    680  106  110       90    130           1   
249    Lugia  Psychic  Flying    680  106   90      130    110           2   
250    Ho-Oh     Fire  Flying    680  106  130       90     90           2   
289  Slaking   Normal     NaN    670  150  160      100    100           3   
382   Kyogre    Water     NaN    670  100  100       90     90           3   

     isLegendary   Color  hasGender  Height_m  Weight_kg  Catch_Rate  \
150         True  Purple      False      2.01      122.0           3   
249         True   White      False      5.21      216.0           3   
250         True     Red      False      3.81      199.0           3   
289        False   Brown       True      2.01      130.5          45   
382         True    Blue      False      4.50      352.0           3   

           Body_Style  Random_status  Att+Def Strong?  
150    bipedal_tailed             79     98.0       Y  
249         two_wings             16    114.0       Y  
250         two_wings             23    106.0       Y  
289  bipedal_tailless             69    124.0       Y  
382         with_fins             38     94.0       Y

<br>

<font color = '#5a86ad'>**Q.pokemon의 Generation이 1세대인 것들만 불러오기**</font>
<br>
Tip: Python의 비교연산자
<br>== : equal
<br>>,<,>=,<= : larger/less than
<br>&,and : and
<br>|,or : or

<font color = '#d9544d'>**Q.pokemon의 Generation이 2세대이면서 동시에 Color가 Green, Red인 것들만 불러오기**</font>

<font color = '#5a86ad'>**Q. 체력(HP)이 150이상인 포켓몬들만 불러오기**</font>

<font color = '#d9544d'>**Q. 이름이 P또는 F로 시작하는 포켓몬의 모든 정보를 불러오기**</font>

Tip: Series에 python string의 method를 쓰고 싶다면? Series.str를 하자!
<br>
Tip: string을 indexing하는 법: string[n] = (n-1)번째 자리의 문자를 return 
<br>
ex: a = "Ybigta"
<br>
print(a[2])
<br>
returns: i

---
# 5. Groupby
---
Groupby는 단순하게, **특정 기준으로 데이터를 분류하는 방법**으로 이해할 수 있습니다.<br>
예를 들어 그 분류기준이 색깔이라면, 색깔을 기준으로 그룹을 나누어서 색에 따른 각 포켓몬들의 특징을 확인할 수 있습니다.<br>

---
#### * 기본적인 사용 방법 : key를 기준으로 묶는다! -> 원하는 column 선택-> operation을 적용


같은 key를 가진 데이터끼리 묶어서
그룹마다 column을 뽑아 내서 (인덱싱) operation을 적용하여 계산한 뒤
각 그룹의 결과를 합쳐서 하나의 데이터프레임으로 만든다.

- <font color="#ff5b00">df.groupby(key)[column].operation()</font>
- 대표적인 operation의 종류
    - size() : 갯수
    - sum() : 합
    - mean() : 평균
    - median() : 중앙값
    - std() : 표준편차
    - max() : 최댓값
    - min() : 최솟값
    - apply() : 내가 만든 함수 적용
    - <font color="#ff5b00">agg(['function1', 'function2', ...])</font> : 여러 개의 함수를 동시에 적용!

pokemon을 [Generation]별로 묶어서 세대별 row의 [size]를 보고 싶다면?

In [70]:
pokemon.groupby('Generation').size()

Generation
1    151
2    100
3    135
4    107
5    156
6     72
dtype: int64

---
#### * grouped : groupby로 묶인 상태에서 데이터 다루기

- <font color="#ff5b00">df.get_group()</font>
- <font color="#ff5b00">filter()</font>
- <font color="#ff5b00">agg()의 활용</font>

In [71]:
grouped = pokemon.groupby("Generation")

In [72]:
# 총 6개의 그룹이 생성된 것을 확인

for name,group in grouped:
    print (type(name))
    print (type(group))
    print (len(group))

<class 'int'>
<class 'pandas.core.frame.DataFrame'>
151
<class 'int'>
<class 'pandas.core.frame.DataFrame'>
100
<class 'int'>
<class 'pandas.core.frame.DataFrame'>
135
<class 'int'>
<class 'pandas.core.frame.DataFrame'>
107
<class 'int'>
<class 'pandas.core.frame.DataFrame'>
156
<class 'int'>
<class 'pandas.core.frame.DataFrame'>
72


In [73]:
# 각 그룹을 그 자체로 불러오는 방법

grouped.get_group(1).head()

Name Type_1  Type_2  Total  HP  ATK  Defense  Speed  Generation  \
1   Bulbasaur  Grass  Poison    318  45   49       49     45           1   
2     Ivysaur  Grass  Poison    405  60   62       63     60           1   
3    Venusaur  Grass  Poison    525  80   60       83     80           1   
4  Charmander   Fire     NaN    309  39   52       43     65           1   
5  Charmeleon   Fire     NaN    405  58   64       58     80           1   

   isLegendary  Color  hasGender  Height_m  Weight_kg  Catch_Rate  \
1        False  Green       True      0.71        6.9          45   
2        False  Green       True      0.99       13.0          45   
3        False  Green       True      2.01      100.0          45   
4        False    Red       True      0.61        8.5          45   
5        False    Red       True      1.09       19.0          45   

       Body_Style  Random_status  Att+Def Strong?  
1       quadruped             95     49.0       N  
2       quadruped             33     62.6       N  
3       quadruped             93     82.6       N  
4  bipedal_tailed             26     46.6       N  
5  bipedal_tailed             39     60.4       N

In [74]:
grouped.filter(lambda x: len(x) <= 120).head()

Name Type_1 Type_2  Total  HP  ATK  Defense  Speed  Generation  \
152  Chikorita  Grass    NaN    318  45   49       65     45           2   
153    Bayleef  Grass    NaN    405  60   62       80     60           2   
154   Meganium  Grass    NaN    525  80   82      100     80           2   
155  Cyndaquil   Fire    NaN    309  39   52       43     65           2   
156    Quilava   Fire    NaN    405  58   64       58     80           2   

     isLegendary   Color  hasGender  Height_m  Weight_kg  Catch_Rate  \
152        False   Green       True      0.89        6.4          45   
153        False   Green       True      1.19       15.8          45   
154        False   Green       True      1.80      100.5          45   
155        False  Yellow       True      0.51        7.9          45   
156        False  Yellow       True      0.89       19.0          45   

           Body_Style  Random_status  Att+Def Strong?  
152         quadruped             67     58.6       N  
153         quadruped             92     72.8       N  
154         quadruped             42     92.8       N  
155  bipedal_tailless             56     46.6       N  
156         quadruped             56     60.4       N

In [75]:
# Total의 max 값이 600이상인 타입의 그룹들만 모아서 보여주기

pokemon.groupby("Type_1").filter(lambda x: x["Total"].max() > 600).head()

Name Type_1  Type_2  Total  HP  ATK  Defense  Speed  Generation  \
4  Charmander   Fire     NaN    309  39   52       43     65           1   
5  Charmeleon   Fire     NaN    405  58   64       58     80           1   
6   Charizard   Fire  Flying    534  78   84       78    100           1   
7    Squirtle  Water     NaN    314  44   48       65     43           1   
8   Wartortle  Water     NaN    405  59   63       80     58           1   

   isLegendary Color  hasGender  Height_m  Weight_kg  Catch_Rate  \
4        False   Red       True      0.61        8.5          45   
5        False   Red       True      1.09       19.0          45   
6        False   Red       True      1.70       90.5          45   
7        False  Blue       True      0.51        9.0          45   
8        False  Blue       True      0.99       22.5          45   

       Body_Style  Random_status  Att+Def Strong?  
4  bipedal_tailed             26     46.6       N  
5  bipedal_tailed             39     60.4       N  
6  bipedal_tailed             77     80.4       N  
7  bipedal_tailed             11     58.2       N  
8  bipedal_tailed             22     73.2       N

In [76]:
# 이런 것까지 할 수 있다 정도만 알아두시면 좋습니다.

grouped = pokemon.groupby(['Type_1', 'isLegendary']).agg({'Height_m': "mean",     
                                     'Type_2': "nunique", 
                                     'Total': ['mean', 'max', 'var']})
grouped

Height_m  Type_2       Total                   
                          mean nunique        mean  max           var
Type_1   isLegendary                                                 
Bug      False        0.859206      11  365.126984  600  12398.015873
Dark     False        0.953846       8  418.961538  600   8829.798462
         True         3.645000       1  640.000000  680   3200.000000
Dragon   False        1.435882       2  443.941176  600  13262.808824
         True         3.502857       6  642.857143  680   1657.142857
Electric False        0.824545       5  406.212121  540   8935.297348
         True         1.670000       1  580.000000  580      0.000000
Fairy    False        0.700000       1  396.500000  552  11300.400000
         True         3.000000       0  680.000000  680           NaN
Fighting False        1.168400       4  404.360000  525   9044.740000
Fire     False        1.032381       6  423.380952  555   8195.997677
         True         2.266000       3  608.000000  680   1720.000000
Flying   False        1.005000       1  390.000000  535  42050.000000
         True         1.500000       0  580.000000  580           NaN
Ghost    False        1.046364       5  412.000000  525   7738.380952
         True         6.910000       1  680.000000  680           NaN
Grass    False        0.980625       9  403.921875  535   8913.914435
         True         1.105000       1  590.000000  600    200.000000
Ground   False        1.115000       8  405.714286  535   8528.656085
         True         2.505000       1  635.000000  670   2450.000000
Ice      False        1.155238       5  412.523810  535   9802.161905
         True         1.750000       1  580.000000  580      0.000000
Normal   False        0.925495       6  385.505495  670  10483.608303
         True         3.455000       0  695.000000  720   1250.000000
Poison   False        1.215000       7  399.142857  535   8530.126984
Psychic  False        0.868205       4  407.615385  600  10847.400810
         True         1.342500       3  612.500000  680   1821.428571
Rock     False        1.315000      11  426.052632  600   7362.267425
         True         1.440000       2  586.666667  600    133.333333
Steel    False        1.398333       6  432.666667  600   8461.529412
         True         2.432500       3  610.000000  680   2266.666667
Water    False        1.255098      14  410.549020  600   9249.438167
         True         3.566667       1  643.333333  680   3033.333333

In [77]:
# 더블 인덱스가 보기싫어서 정리할 때 -> droplevel

grouped.columns = grouped.columns.droplevel(level=0)
grouped

mean  nunique        mean  max           var
Type_1   isLegendary                                                  
Bug      False        0.859206       11  365.126984  600  12398.015873
Dark     False        0.953846        8  418.961538  600   8829.798462
         True         3.645000        1  640.000000  680   3200.000000
Dragon   False        1.435882        2  443.941176  600  13262.808824
         True         3.502857        6  642.857143  680   1657.142857
Electric False        0.824545        5  406.212121  540   8935.297348
         True         1.670000        1  580.000000  580      0.000000
Fairy    False        0.700000        1  396.500000  552  11300.400000
         True         3.000000        0  680.000000  680           NaN
Fighting False        1.168400        4  404.360000  525   9044.740000
Fire     False        1.032381        6  423.380952  555   8195.997677
         True         2.266000        3  608.000000  680   1720.000000
Flying   False        1.005000        1  390.000000  535  42050.000000
         True         1.500000        0  580.000000  580           NaN
Ghost    False        1.046364        5  412.000000  525   7738.380952
         True         6.910000        1  680.000000  680           NaN
Grass    False        0.980625        9  403.921875  535   8913.914435
         True         1.105000        1  590.000000  600    200.000000
Ground   False        1.115000        8  405.714286  535   8528.656085
         True         2.505000        1  635.000000  670   2450.000000
Ice      False        1.155238        5  412.523810  535   9802.161905
         True         1.750000        1  580.000000  580      0.000000
Normal   False        0.925495        6  385.505495  670  10483.608303
         True         3.455000        0  695.000000  720   1250.000000
Poison   False        1.215000        7  399.142857  535   8530.126984
Psychic  False        0.868205        4  407.615385  600  10847.400810
         True         1.342500        3  612.500000  680   1821.428571
Rock     False        1.315000       11  426.052632  600   7362.267425
         True         1.440000        2  586.666667  600    133.333333
Steel    False        1.398333        6  432.666667  600   8461.529412
         True         2.432500        3  610.000000  680   2266.666667
Water    False        1.255098       14  410.549020  600   9249.438167
         True         3.566667        1  643.333333  680   3033.333333

<br>
<br>
<br>
<font color = '#5a86ad'>**Q. pokemon을 Color를 기준으로 묶어서 Total의 평균을 보고 싶다면?**</font>

<font color = '#d9544d'>**Q. pokemon을 Color와 isLegendary 별로 나누었을 때 Attack의 최댓값은?**</font>

<font color = '#d9544d'>**Q. pokemon에서 1세대 포켓몬 중 이름이 가장 뒤에 있는 포켓몬과 2세대 중 이름이 가장 뒤에있는 포켓몬?**</font>

<font color = '#77926f'> **Tip -  이처럼 조건문과 groupby를 잘 쓸 줄 알아야 변수를 쉽게 가공할 수 있습니다. 나중에 배울 데이터 시각화에서도 중요한 이슈가 되죠. 그리고 위의 문제처럼 창의적인 파생변수를 만들고자 한다면 더더욱 잘 쓸 줄 알아야 해요...** </font>